core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java

```diff
@@ -143,27 +145,29 @@ public List<TableIdentifier> listTables(Namespace namespace) {
   }
 
   @Override
-  public synchronized void renameTable(TableIdentifier from, TableIdentifier to) {
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
     if (from.equals(to)) {
       return;
     }
 
-    if (!namespaceExists(to.namespace())) {
-      throw new NoSuchNamespaceException(
-          "Cannot rename %s to %s. Namespace does not exist: %s", from, to, to.namespace());
-    }
+    synchronized (this) {
+      if (!namespaceExists(to.namespace())) {
```

Comment:
I think we want to include the namespace existence check inside the sync block, as otherwise it could be possible that the namespace is concurrently dropped